### PR TITLE
8708 Update to support Rails 4.2

### PR DESCRIPTION
**Pull request by baburdick**
Task 8708

This is preliminary work to support Rails 4.2.

### CODE REVIEW

**Comment by baburdick** [comment]
When these gem updates are applied, the following spec failures appear:
* arel 6.0.3
* activesupport 4.2.6
* activemodel 4.2.6
* activerecord 4.2.6
```
Failures:

  1) Skipping #use! should enable skipped mounting
     Failure/Error: super
     
     NoMethodError:
       undefined method `human' for "mapper":String
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:437:in `generate_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:449:in `normalize_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:300:in `add'
     # ./lib/flat_map/errors.rb:54:in `add'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations/numericality.rb:33:in `validate_each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:151:in `block in validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `public_send'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `block in make_lambda'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `block in simple'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `block in call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:778:in `_run_validate_callbacks'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:399:in `run_validations!'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:338:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:125:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `map'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./spec/flat_map/mapper/skipping_spec.rb:40:in `block (2 levels) in <module:FlatMap>'

  2) Validations should call callbacks
     Failure/Error: super
     
     NoMethodError:
       undefined method `human' for "mapper":String
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:437:in `generate_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:449:in `normalize_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:300:in `add'
     # ./lib/flat_map/errors.rb:54:in `add'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations/presence.rb:7:in `validate_each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:151:in `block in validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `public_send'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `block in make_lambda'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `block in simple'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `block in call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:778:in `_run_validate_callbacks'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:399:in `run_validations!'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:338:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:125:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `map'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./spec/flat_map/mapper/validations_spec.rb:60:in `block (2 levels) in <module:FlatMap>'

  3) Validations should not be valid
     Failure/Error: super
     
     NoMethodError:
       undefined method `human' for "mapper":String
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:437:in `generate_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:449:in `normalize_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:300:in `add'
     # ./lib/flat_map/errors.rb:54:in `add'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations/presence.rb:7:in `validate_each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:151:in `block in validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `public_send'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `block in make_lambda'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `block in simple'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `block in call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:778:in `_run_validate_callbacks'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:399:in `run_validations!'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:338:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:125:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `map'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./spec/flat_map/mapper/validations_spec.rb:55:in `block (2 levels) in <module:FlatMap>'

  4) Validations should have all the errors
     Failure/Error: super
     
     NoMethodError:
       undefined method `human' for "mapper":String
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:437:in `generate_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:449:in `normalize_message'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/errors.rb:300:in `add'
     # ./lib/flat_map/errors.rb:54:in `add'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations/presence.rb:7:in `validate_each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:151:in `block in validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validator.rb:148:in `validate'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `public_send'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:455:in `block in make_lambda'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:192:in `block in simple'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `block in call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `each'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:504:in `call'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activesupport-4.2.6/lib/active_support/callbacks.rb:778:in `_run_validate_callbacks'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:399:in `run_validations!'
     # /Users/bruceb/.rvm/gems/ruby-2.2.4@flat_map/gems/activemodel-4.2.6/lib/active_model/validations.rb:338:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:125:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `map'
     # ./lib/flat_map/open_mapper/persistence.rb:124:in `valid?'
     # ./lib/flat_map/open_mapper/skipping.rb:40:in `valid?'
     # ./spec/flat_map/mapper/validations_spec.rb:65:in `block (2 levels) in <module:FlatMap>'

Finished in 0.30818 seconds (files took 0.62098 seconds to load)
164 examples, 4 failures

Failed examples:

rspec ./spec/flat_map/mapper/skipping_spec.rb:37 # Skipping #use! should enable skipped mounting
rspec ./spec/flat_map/mapper/validations_spec.rb:58 # Validations should call callbacks
rspec ./spec/flat_map/mapper/validations_spec.rb:54 # Validations should not be valid
rspec ./spec/flat_map/mapper/validations_spec.rb:64 # Validations should have all the errors

Randomized with seed 55245
```

**Comment by baburdick** [comment]
[This change](https://github.com/rails/rails/commit/6b0e834a194d112585f41deba0a40780d68e38c6#diff-fdcf8b65b5fb954372c6fe1ddf284c78) is essential to understanding the problem.

* In Rails 4.1.15, `ActiveModel::Errors#generate_message` uses this expression to get the model name of the mapper:
** `@base.class.model_name.human` retrieving for example this value: "Spec mapper with trait trait"
* In Rails 4.2.6, `ActiveModel::Errors#generate_message` uses this expression to get the model name of the mapper:
** `@base.model_name.human` Generating the error above.

* `@base.model_name` has this value: "mapper"
* `@base.class.model_name` has this value:

```
#<ActiveModel::Name:0x007f93c0b81d10
 @collection="flat_map/skipping_spec/spec_mapper_with_trait_traits",
 @element="spec_mapper_with_trait_trait",
 @human="Spec mapper with trait trait",
 @i18n_key=:"flat_map/skipping_spec/spec_mapper_with_trait_trait",
 @klass=#<Class:0x007f93bc5722e8>,
 @name="FlatMap::SkippingSpec::SpecMapperWithTraitTrait",
 @param_key="flat_map_skipping_spec_spec_mapper_with_trait_trait",
 @plural="flat_map_skipping_spec_spec_mapper_with_trait_traits",
 @route_key="flat_map_skipping_spec_spec_mapper_with_trait_traits",
 @singular="flat_map_skipping_spec_spec_mapper_with_trait_trait",
 @singular_route_key="flat_map_skipping_spec_spec_mapper_with_trait_trait">
```

So FlatMap must be adjusted in the context of Rails 4.2 so that `@base.model_name` returns an `ActiveModel::Name` object instead of a string. Or perhaps deeper/more extensive changes are needed.

**Comment by baburdick** [comment]
This change here appears to have set the stage for this issue: https://github.com/HornsAndHooves/flat_map/commit/bf5c7f78845f831bad7d0f0300844fad59a1c0cb#diff-44b699ac27c73c58f925c41ea7d427abR57

**Comment by baburdick** [comment]
c0b6348 fixes specs. But I'm not certain this won't cause issues with `FormBuilder` that are not covered in the test suite. @albertosaurus, @avolosenko, @DShein, @akuzko: Thoughts?

File: HornsAndHooves-flat_map.gemspec [hunk 1]
```diff
@@ -19,8 +19,8 @@ Gem::Specification.new do |s|
   s.require_paths = ["lib"]
 
   # specify any dependencies here; for example:
-  s.add_dependency(%q<activesupport>, ["> 4.0", "< 4.2"])
-  s.add_dependency(%q<activerecord>, ["> 4.0", "< 4.2"])
+  s.add_dependency(%q<activesupport>, ["> 4.0", "< 5"])
```
Comment (albertosaurus) [inline HornsAndHooves-flat_map.gemspec]: This needs to be `~> 4.2.0`

**Comment by albertosaurus** [comment]
I actually think it'll work. Need to test and confirm.

File: HornsAndHooves-flat_map.gemspec (hunk 1 above)
Comment (baburdick) [inline HornsAndHooves-flat_map.gemspec]: With Rails 5 coming out, there won't be a Rails 4.3. And this still supports Rails 4.1.

Comment (albertosaurus) [inline HornsAndHooves-flat_map.gemspec]: OK.

**Comment by albertosaurus** [comment]
@baburdick we should merge and release this.

**Pull request merged**